### ISSUE TITLE
Provide Meteor.userId() inside a publish function

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#5462]
**_2 Upvotes**_ There is really no reason why `Meteor.userId()` could not be provided inside a publish function. The comment in the code:

```
// This function only works if called inside a method. In theory, it
// could also be called from publish statements, since they also
// have a userId associated with them. However, given that publish
// functions aren't reactive, using any of the infomation from
// Meteor.user() in a publish function will always use the value
// from when the function first runs. This is likely not what the
// user expects. The way to make this work in a publish is to do
// Meteor.find(this.userId()).observe and recompute when the user
// record changes.
```

Is invalid because publish function is anyway rerun every time `userId` changes. So `Meteor.userId()` could be available (so that one can reuse the code between publish functions and methods). I agree that `Meteor.user()` might throw an error then instead. (But even that is not true anymore with [reactive-publish](https://github.com/peerlibrary/meteor-reactive-publish).)

So, check for publish function should be moved to `Meteor.user()` at least and `Meteor.userId()` should be made so that it works in the publish function as well.

**Comment by stubailo** [comment]
That makes sense to me; I wonder if there are any edge cases where `this.userId` would do something more reasonable than `Meteor.userId()`. You could run into some weird cases with asynchronous code, maybe?

**Comment by mitar** [comment]
So currently I am doing:

```coffee
# Override the publish function so that we can get the current userId.
publishCurrentContext = new Meteor.EnvironmentVariable()

originalPublish = Meteor.publish
Meteor.publish = (name, func) ->
  originalPublish name, (args...) ->
    publishCurrentContext.withValue {userId: @userId}, =>
      func.apply @, args

Meteor.userId = ->
  currentInvocation = DDP._CurrentInvocation.get()

  return currentInvocation.userId if currentInvocation

  currentContext = publishCurrentContext.get()

  return currentContext.userId if currentContext

  throw new Error "Meteor.userId() not invoked from a method or publish function."
```

So `Meteor.userId()` is exactly the same as `this.userId` when called inside a `bindEnironment` callbacks. Outside, yea, things will work less well, but then you can still go and use `this.userId`.

**Comment by matteodem** [comment]
+1

I'm currently also running into this issue where I need to provide a consistent approach to retrieving the userId in different environments. I'll take mitars code for now but this should be part of core at some time.

**Comment by mitar** [comment]
I made a package for this: https://github.com/peerlibrary/meteor-user-extra

**Comment by mitar** [comment]
#5462 is probably a duplicate of this ticket.

**Comment by avalanche1** [comment]
+1

**Comment by mjmasn** [comment]
Know this issue is pretty old but on a similar vein of thought - although I'm not that interested in `Meteor.userId()` - I'm trying to be able to call `DDP._currentInvocation.get()` inside a publish function (and in any function that is called from that publish function) because reasons*

I made some changes to `livedata_server.js` - basically copying the invocation stuff from the `method` handler into the `sub` handler:

It appears to work fine, but I'm sure there's a reason this wasn't done... Anybody know if doing this is a really bad idea? Performance? Security?

```js
  protocol_handlers: {
    sub: function (msg, unblock) {

      // ...snip...

      var randomSeed = msg.randomSeed || null;

      var setUserId = function(userId) {
        self._setUserId(userId);
      };

      var invocation = new DDPCommon.MethodInvocation({
        isSimulation: false,
        userId: self.userId,
        setUserId: setUserId,
        unblock: unblock,
        connection: self.connectionHandle,
        randomSeed: randomSeed
      });

      DDP._CurrentInvocation.withValue(
        invocation, function () {
          self._startSubscription(handler, msg.id, msg.params, msg.name);
        }
      );
  }
```

*I want to easily access the connectionId for audit logging within our system, userId isn't enough because a user can have multiple sessions or they may be being impersonated by another user. Saving the connectionId helps us tie everything together (because we can also log the type of login for a connectionId) and it works great for methods, but not publications. I know the publication has access to `this.connection` but then that would have to be passed as a parameter to various other layers of functions which is not ideal.

**Comment by abernix** [comment]
I think #8031 is suggesting the implementation of the opposite of what you want, but perhaps your input might be helpful.

**Comment by mjmasn** [comment]
I just want it to be easy to obtain the connection details for any piece of code* running on behalf of that connection (basically subs and methods, including logins). Maybe not the most common use case but if you're doing anything that requires logging of who did what and when it's suddenly very necessary. I haven't had loads of time to think about this so there may be better ways of achieving that goal, I just wanted to avoid passing extra parameters everywhere - but that is probably a more 'correct' and testable solution than globals and meteor magic.

* anything synchronous anyway, things are way less clear once you get into async stuff...

**Comment by mitar** [comment]
[publish-context](https://github.com/peerlibrary/meteor-publish-context) package provides you now with whole publish context in code called from a publish function.

**Comment by hwillson** [comment]
To help provide a more clear separation between feature requests and bugs, and to help clean up the feature request backlog, Meteor feature requests are now being managed under the https://github.com/meteor/meteor-feature-requests repository. 

This feature request will be closed here, but anyone interested in migrating this feature request to the new repository (to make sure it stays active), can click [here](https://github.com/meteor/meteor-feature-requests/issues/new?title=Provide%20Meteor.userId()%20inside%20a%20publish%20function&body=Migrated%20from:%20meteor/meteor%235462) to start the feature request migration process. This manual migration process is intended to help identify which of the older feature requests are still considered to be of value to the community. Thanks!

**Comment by mitar** [comment]
I think this was resolved with #8629.
